README.md
# imapfilter

> IMAP message filtering and actions for humans, in pure Haskell

## Table of Contents

- [About](#about)
- [Getting Started](#getting-started)
- [Usage](#usage)
- [Contributing](../CONTRIBUTING.md)

## About

A library that filters messages found using human-readable search queries that
resemble Gmail operations. The command-line tool leverages that to act upon the
search results.

The current version implements the expression interface for the [HaskellNet]
`SearchQuery` type.

NOTE: This is a work-in-progress, use at your own risk! 😷

## Getting Started

The project uses Stack, so you can clone and `stack build` to get it up and
running locally.

## Usage

Expressions can contain either literal terms to search as plain text and/or
operators to configure specific criteria. The query language supports all
[RFC 9051 search keys][rfc-9051-search-command], plus some
[Gmail-specific operators][gmail-operators] for compatibility. Here's a
non-exhaustive list of criteria supported and how to use it:

[rfc-9051-search-command]: https://datatracker.ietf.org/doc/html/rfc9051#name-search-command
[gmail-operators]: https://support.google.com/mail/answer/7190?hl=en

| Criteria                | Query                                         |
|-------------------------|-----------------------------------------------|
| `FROM/TO/CC/BCC`        | `from:foo@bar.com`                            |
| `DELETED/DRAFT/FLAGGED` | `is:draft`, `-is:flagged` (or `is:unflagged`) |
| `ANSWERED/SEEN/NEW/OLD` | `is:answered`, `-is:seen` (or `is:unseen`)    |
| `KEYWORD`               | `keyword:foo`                                 |
| `SUBJECT/BODY`          | `subject:lorem`, `subject:(lorem ipsum)`      |
| `BEFORE/ON/SINCE`       | `before:2023/4/18`                            |
| `SENTBEFORE/-ON/-SINCE` | `sent_before:2023/4/18`                       |
| `HEADER`                | `header:foo=bar`                              |
| `LARGER/SMALLER`        | `larger:10`                                   |
| `NOT`                   | `-from:qux@bar.com`                           |
| `OR`                    | `{from:foo@bar.com from:qux@bar.com}`         |

Gmail-compatibility ones:

| Criteria      | Query                                |
|---------------|--------------------------------------|
| `SEEN/UNSEEN` | `is:read`, `is:unread`               |
| `LARGER`      | `size:10`                            |
| `SINCE`       | `after:2023/4/18`, `newer:2023/4/18` |
| `BEFORE`      | `older:2023/4/18`                    |

Check the `examples` folder in this repository for samples on how to use the
expressions directly in code or with common libraries such as `HaskellNet` and
`imap`.

### inline queries

You can build library-independent queries directly in code using the
`SearchExpr` class. For that, it's wise to import the module qualified, as some
functions conflict with Prelude, such as `or`, `all` and `not`. For example:

```haskell
import           System.IO.Unsafe    (unsafePerformIO)
import qualified System.Time         as ST
import qualified Network.IMAP.Search as Search (SearchExpr (..))

{-# NOINLINE now #-}
now :: ST.ClockTime
now = unsafePerformIO ST.getClockTime

1dago :: ST.ClockTime
1dago = ST.addToClockTime (ST.TimeDiff {ST.tdDay = -1}) now

read1dago :: SearchExpr a => [a]
read1dago = [ Search.seen
            , Search.before (ST.toCalendarTime 1dago)
            ]
```

You can then use `read1dago` with any library interface that takes a list of
query operations, such as `HaskellNet`'s `search`. You can also transform it to
the generic `Operation` type provided, which you can transform to string or text
for libraries that don't have proper typing for the IMAP search keys:

```haskell
```

### With HaskellNet

```haskell
module Main (main) where

import Control.Monad      (forM_)
import System.Environment (getArgs, getEnv)

import Network.HaskellNet.IMAP (fetchHeaderFields, login, search, select)

import Network.HaskellNet.IMAP.Connection (IMAPConnection)
import Network.HaskellNet.IMAP.Search     (parseSearchQuery)
import Network.HaskellNet.IMAP.SSL        (connectIMAPSSL)

import qualified Data.ByteString as BS

main :: IO ()
main = do
  query <- either (error . show) id . parseSearchQuery . unwords <$> getArgs

  putStr "parsed query: "
  print query

  con <- connectWithEnv
  loginWithEnv con

  select con "INBOX"

  messages <- search con query
  putStrLn "results:"
  separatorLn
  forM_ messages $ \m -> do
    headers <- fetchHeaderFields con m ["FROM", "TO", "SUBJECT"]
    BS.putStr headers
    separatorLn

connectWithEnv :: IO IMAPConnection
connectWithEnv = getEnv "IMAP_SERVER" >>= connectIMAPSSL

loginWithEnv :: IMAPConnection -> IO ()
loginWithEnv conn = do
  username <- getEnv "IMAP_USERNAME"
  password <- getEnv "IMAP_PASSWORD"
  login conn username password

separatorLn :: IO ()
separatorLn = putStrLn $ replicate 80 '#'
```


CHANGELOG.md
# Changelog for `imapfilter`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Network/IMAP/Search/Internal/Parser.hs
module Network.IMAP.Search.Internal.Parser (
  parseSearch
) where

import Control.Monad (void)
import Data.Functor  (($>))

import Text.Parsec

import Data.Time (TimeLocale, ZonedTime, defaultTimeLocale, parseTimeOrError)

import           Network.IMAP.Search.Internal.SearchExpr
import qualified Network.IMAP.Search.Internal.SearchExpr as Search (SearchExpr (..))

type Parser a = Parsec String () a

parseSearch :: SearchExpr q => String -> Either ParseError [q]
parseSearch = parse query ""

query :: SearchExpr q => Parser [q]
query = spaces *> sepEndBy operation (spaces <|> void endOfLine)

operation :: SearchExpr q => Parser q
operation = spaces *> choice operations <|> (text <$> literal)
  where
    operations :: SearchExpr q => [Parser q]
    operations =
      [ Search.all <$ static "in:all" -- for completeness sake :P
      , answered <$ static "is:answered"
      , bcc <$> prefixed "bcc:" address
      , before <$> prefixed "before:" calendarTime
      , body <$> prefixed "body:" literal
      , cc <$> prefixed "cc:" address
      , deleted <$ static "is:deleted"
      , draft <$ static "is:draft"
      , flagged <$ static "is:flagged"
      , from <$> prefixed "from:" address
      , header <$> prefixed "header:" fieldName <*> literal
      , keyword <$> prefixed "keyword:" literal
      , larger <$> prefixed "larger:" digits
      , new <$ static "is:new"
      , old <$ static "is:old"
      , on <$> prefixed "on:" calendarTime
      , braces (Search.or <$> operation <*> operation)
      -- TODO parenthesized or with keyword
      , recent <$ static "is:recent"
      , seen <$ static "is:seen"
      , sentBefore <$> prefixed "sent_before:" calendarTime
      , sentOn <$> prefixed "sent_on:" calendarTime
      , sentSince <$> prefixed "sent_since:" calendarTime
      , since <$> prefixed "since:" calendarTime
      , smaller <$> prefixed "smaller:" digits
      , subject <$> prefixed "subject:" literal
      , to <$> prefixed "to:" address
      -- TODO uid
      , unanswered <$ static "is:unanswered"
      , unanswered <$ static "-is:answered"
      , undeleted <$ static "is:undeleted"
      , undeleted <$ static "-is:deleted"
      , undraft <$ static "is:undraft"
      , undraft <$ static "-is:draft"
      , unflagged <$ static "is:unflagged"
      , unflagged <$ static "-is:flagged"
      , unkeyword <$> prefixed "unkeyword:" literal
      , unkeyword <$> prefixed "-keyword:" literal
      , unseen <$ static "is:unseen"
      , unseen <$ static "-is:seen"

      -- NOT must be last as there's static parsers prefixed with minus as well
      , Search.not <$> prefixed "-" operation
      , Search.not . text <$> prefixed "-" literal

      -- Gmail operators compatibility
      , Search.all <$ (static "in:anywhere" *> parserFail "IMAP search works only within a mailbox")
      , seen <$ static "is:read"
      , unseen <$ static "is:unread"
      , since <$> prefixed "after:" calendarTime
      , before <$> prefixed "before:" calendarTime
      , before <$> prefixed "older:" calendarTime
      , since <$> prefixed "newer:" calendarTime
      , larger <$> prefixed "size:" digits
      , text <$> prefixed "+" literal
      ]

braces :: Parser p -> Parser p
braces = between (spaces *> char '{') (spaces *> char '}')

prefixed :: String -> Parser p -> Parser p
prefixed a b = try (string a) *> b

-- it is not our objective to fully validate email addresses
address :: Parser String
address = many1 (alphaNum <|> oneOf ".@-+_")

static :: String -> Parser ()
static p = try (string p) $> ()

digits :: (Read a, Num a) => Parser a
digits = read <$> many1 digit

-- FIX parse returns UTC instead of TZ-aware time
calendarTime :: Parser ZonedTime
calendarTime = choice
  [ parsecZonedTime defaultTimeLocale "%Y/%m/%d"
  , parsecZonedTime defaultTimeLocale "%Y/%m/%d"
  ]

parsecZonedTime :: TimeLocale -> String -> Parser ZonedTime
parsecZonedTime t s = try (parseTimeOrError False t s <$> many1 (digit <|> char '/'))

-- oldCalendarTime :: Parser OST.CalendarTime
-- oldCalendarTime = choice
--   [ OSTP.parsecCalendarTime OSL.defaultTimeLocale "%Y/%m/%d"
--   , OSTP.parsecCalendarTime OSL.defaultTimeLocale "%m/%d/%Y" -- BEWARE: this is mostly to support Gmail
--   ]

literal :: Parser String
literal = doubleQuoted <|> singleQuoted <|> parenthesized <|> word
  where
    word = many1 (noneOf " (){}\r\n")
    doubleQuoted = between (char '"') (char '"') (many (noneOf "\"\r\n"))
    singleQuoted = between (char '\'') (char '\'') (many (noneOf "'\r\n"))
    parenthesized = between (char '(') (char ')') (many (noneOf ")\r\n"))

fieldName :: Parser String
fieldName = manyTill anyChar (try $ char '=')

-- RFC 3501 - IMAP 4rev1
-- [ ] <sequence set>
-- [x] all
-- [x] answered
-- [x] bcc
-- [x] before
-- [x] body
-- [x] cc
-- [x] deleted
-- [x] draft
-- [x] flagged
-- [ ] or
--   [x] {}
--   [ ] OR in-between
-- [x] from (plain)
-- [x] header
-- [x] keyword
-- [x] larger
-- [x] new
-- [x] not
-- [x] old
-- [x] on
-- [x] recent
-- [x] seen
-- [x] sent before
-- [x] sent on
-- [x] sent since
-- [x] since
-- [x] smaller
-- [x] subject
-- [x] text
-- [x] to
-- [ ] uid <sequence set>
-- [x] unanswered
-- [x] undeleted
-- [x] undraft
-- [x] unflagged
-- [x] unkeyword
-- [x] unseen

-- RFC 9051 - IMAP 4rev2
-- https://datatracker.ietf.org/doc/html/rfc9051
-- [ ] SAVE support

-- Gmail search operators https://support.google.com/mail/answer/7190?hl=en
-- they can be used also with the custom X-GM-RAW search attribute
-- [ ] larger/smaller/size
--   [x] plain bytes
--   [ ] with units
-- [ ] OR/{}/inline e.g. from:(foo@quux.com OR bar@quux.com)
-- [x] - (negate)
-- [ ] AROUND
-- [x] has (<custom flag>)
-- [ ] list
-- [ ] filename
-- [x] " " (exact search)
-- [x] is:unread
-- [x] is:read
-- [x] after/newer/before/older
-- [x] + (exact word)
-- [ ] older_than/newer_than
-- [ ] deliveredto
-- [ ] Rfc822msgid
-- [ ] category (maybe general if using openXchange categories?)
-- [x] size

-- Gmail IMAP extensions https://developers.google.com/gmail/imap/imap-extensions
-- TODO parse the capability X-GM-EXT-1
-- [ ] has:drive
-- [ ] has:document
-- [ ] has:spreadsheet
-- [ ] has:presentation
-- [ ] has:youtube
-- [ ] has:attachment
-- [ ] has:nouserlabels
-- [ ] has:userlabels
-- [ ] is:important
-- [ ] is:snoozed
-- [ ] is:starred
-- [ ] label
-- [x] in:anywhere (nope, IMAP works within a mailbox)
